% document formatting
\documentclass[10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[left=1in,right=1in,top=1in,bottom=1in]{geometry}
\usepackage[T1]{fontenc}
\usepackage{xcolor}

% math symbols, etc.
\usepackage{amsmath, amsfonts, amssymb, amsthm}

% lists
\usepackage{enumerate}
\usepackage{enumitem}

% images
\usepackage{graphicx} % for images
\usepackage{tikz} % draw stuff

% code blocks
\usepackage{minted, listings}

% links
\usepackage{hyperref}

% \mathbb{1} symbol
\usepackage{dsfont}

\usepackage{tabularx}
\newcolumntype{Y}{>{\centering\arraybackslash}X}

\graphicspath{{./assets/images}}

\newcommand{\solution}{\textbf{Solution:}} 

\title{COM SCI C121 Week 9}

\author{Aidan Jan}
\date{\today}

\begin{document}
\maketitle

\section*{Experimental Design, Robustness, and Functional Genomics}
\subsection*{Functional Genomics}
\begin{itemize}
    \item The study of the function of many genes at once (or how genes bbehave in different contexts based on some form of a perturbation)
    \item "Functional genomics focuses on the \textbf{dynamic expression of gene products in a specific context}, for example, at a specific developmental stage or during a disease.  In functional genomics, we try to use our current knowledge of gene function to develop a model linking genotype to phenotype."
\end{itemize}

\subsection*{Robustness}
\begin{itemize}
    \item Let's assume we have a \textit{hypothesis} we want to ask:
    \begin{itemize}
        \item If I perturb gene $X$, how will gene $Y$ behave?
    \end{itemize}
    \item There is \textit{reality}, there is my belief of reality, and there is my tractable representation of my belief of reality
    \begin{itemize}
        \item Which is my generative model and which is my inference model?
    \end{itemize}
    \item Robustness refers to a model that can behave well when there are varying levels of misspecification.
\end{itemize}
\subsubsection*{Experimental design in genomics is hard because of high-dimensional sampels and so many hypotheses}
\begin{itemize}
    \item There is a lot of work in the classical statistical literature around "experimental design"
    \begin{itemize}
        \item e.g., I'm designing a randomized controlled trial to test the efficacy of a drug.  If I believe the effect to be "around" some value, how many samples do I need?
    \end{itemize}
    \item In modern genomics, there are often many moving part in \textit{high dimensions}.
    \item The fact that so many parts are moving makes this question of sample size very difficult because in some sense, every hypothesis is interrelated.
\end{itemize}
\subsubsection*{Experimental design can be used for robustness testing and for designing experiments}
\begin{itemize}
    \item Imagine you have a reasonable generative model of some experimental process
    \item If I want to perform an experiment in a new setting, I can simulate under that process
    \item Because I know what I changed in the generative model, I can evaluate how well I am doing on the inference side.  If the generative model is already different than the inference model, I've done a robustness analysis for free
\end{itemize}

\section*{Functional Genomics Today: Genetic Variation}
\begin{itemize}
    \item A fair amount of work is done on the statistical genetics side for functional genomics
    \item Here, a perturbation is a genetic variant
    \item A simple example is a cis-eQTL
    \begin{itemize}
        \item "cis" = jargon for "nearby".  In practice, within 100000 bases of a gene.
        \item "eQTL" = expression quantitative trait loci.  Jargon for gene expression that changes as a function of the genetic variant.
    \end{itemize}
\end{itemize}

\subsection*{Experimentally Induced Perturbations}
\begin{itemize}
    \item In the past, you pipette some lead into some cells on a dish, then look at gene expression differences
    \item The function transcriptional (gene expression) changes in response to a stress (lead) is an experimentally induced perturbation.
\end{itemize}

\section*{CRISPR}
\begin{itemize}
    \item CRISPR can be used to knock out genes by introducing variants
    \item Briefly, a short guide RNA (sgRNA) \~20 bases long is engineered.
    \item Together, with Cas9, the complex searches for that sequence, cuts the DNA, then the cell "repairs" the DNA, and totally screws up super often, thus breaking the cell (NHEJ, non-homologous end joining)
\end{itemize}

\subsection*{CRISPR Works to change one gene, but how about many genes?}\
\begin{itemize}
    \item Remember engineered barcodes?  We can engineer guides!
    \item There is an entire field of how to generate these efficiently
    \item The GeCKO library targets 19050 genes with 123411 sequences.
    \item It only costs \$600.
\end{itemize}

\subsection*{How do you put CRISPR library DNA into cells}
\begin{itemize}
    \item Amplify the library DNA
    \item Put the library DNA in phages and have them "infect" cells.
    \item Each cell gets a different snippet of DNA and you have basically run 10k experiments in one.
\end{itemize}

\subsection*{Exercise: A Generative Model for Single-Cell Perturbations}
\begin{itemize}
    \item How would you play the oracle for generating single-cell RNA data with perturbations?
\end{itemize}
\begin{enumerate}
    \item Choose a cell type
    \item Choose a gene to knock out
    \item Measure sampling (DNA)
    \item Repeat for every cell in sample.
\end{enumerate}

Cell Types:
\begin{enumerate}
    \item $t_c \sim \text{Categorical}(p_t)$
    \item $k_c \sim \text{Categorical}(p_k)$
    \item for each $g$ in numGenes:
    \begin{itemize}
        \item $y_{cg} \sim \text{Poisson}(\mu_g \exp(\sum_a \alpha_{ga}^{\mathds{1}\{k_i = 1\}}))$
    \end{itemize}
\end{enumerate}
\begin{itemize}
    \item where $p_t$ is the proportion of cell types.
    \item A Categorical distribution is the same as $\text{Multinomial}(1, p_t)$
    \item Part 3 essentially means, what is the probability that the gene is being affected by the given knockout and not another?
    \item $\mu$ is the normal expression rate of a trait (some trait of some cell in standard conditions).  $\alpha$ is the "effect size", a constant.  Positive means the gene being knocked out causes expression to occur more than the mean. Negative means it expresses less than the mean.  Zero means that the removal of that gene does not affect the cell (e.g., expression of the studied trait is the same as the mean).
    \item The goal is to use the knockout and effect data to predict which parts of genes are the most 'important' in what traits the cells express.
\end{itemize}

\section*{Actual Functional Genomics: FACS Screens}
\begin{center}
    \includegraphics*[scale=0.8]{W9_1.png}
\end{center}
\begin{itemize}
    \item Downstream mapping is studying what a changing a gene causes.
    \item Upstream mapping is studying what causes a change in a gene.
\end{itemize}
Most of the time, we want to know the changes in other genes that causes changes in the trait we are observing, but that is really hard to do since we would need to know what all the genes do.

\subsection*{Upstream Regulators can be Inferred by Perturbations}
\begin{center}
    \includegraphics*[scale=0.8]{W9_2.png}
\end{center}
Suppose we care about the trait.  We do a "do() operation" on gene B.  Basically, we set the expression of Gene B to 0 by knocking out the gene. (\texttt{do(Gene B = 0)}).  Now, the only genes affecting the trait are genes A, C, and D, in a straight line.  Now, if we modify Gene C, we can see its effect on the trait.\\\\
We can use this technique, for say, finding genetic variants associated with size of watermelons.
\begin{enumerate}
    \item Take the seeds from one watermelon and mutate the seeds
    \item Grow watermelons and sort them by size.
    \item Analyze DNA of small melons vs big melons.
\end{enumerate}
\begin{center}
    \includegraphics*[scale=0.8]{W9_3.png}
\end{center}
In this example, watermelon size was determined by which segments appeared more often (likely correlates to the amount of certain proteins synthesized.)

\subsection*{Forward Genetics Using FACS-Based Screens Finds Upstream Regulators of a Gene}
\begin{center} 
    \includegraphics*[scale=0.8]{W9_4.png}
\end{center}
\begin{enumerate}
    \item Apply sgRNA and viruses to create perturbed cells.  In FACS, genes are tagged with fluorescent molecules, the color of which is a "barcode" for the gene received.
    \item Let the cells grow and observe the expression of the trait of interest.
    \item Group all the traits into "bins"
    \item Use a light detector on each bin to determine a correlation between the sets of genes (color) with the amount of expression.
\end{enumerate}
\begin{center}
    \includegraphics*[scale=0.8]{W9_5.png}
\end{center}

\subsection*{How Do We Design Better Screens?}
A big problem with flow cytometry is that you need a lot of cells.  There will be read errors, dead cells, and a lot of variation, so to get a good dataset, millions of cells are needed.\\\\
The \textbf{goal} is to design a screen such that we have high confidence in calls and probe as many targets as possible\\\\
Approach:
\begin{enumerate}
    \item Develop a model
    \item Learn parameters from existing data
    \item Simulat unobserved conditions
    \item Evaluate performance
    \item Repeat
\end{enumerate}

\subsection*{Pre-data Modeling Enables Efficient Screen Design...}
... by using prior data for informative predictions
\begin{center}
    \includegraphics*[scale=0.8]{W9_6.png}
\end{center}
This is a generative model.
\begin{itemize}
    \item $M_{nc} \sim \text{Poisson}(\lambda)$
    \item We pick a (single) cell with some number of guides in it.
    \item The $G_{nc}$ node picks which guides they are, where how much each guide affects the observed trait is drawn from $p$, the total proportion of gene effects.  $G_{nc} \sim \text{Mult}(M_{nc}, p)$
    \begin{itemize}
        \item $p \sim \text{Dirichlet}(\frac{1}{N_g} \mathds{1}_{N_g}\delta)$
    \end{itemize}
    \item Skip $D_{nc}$ for now, since it is somewhat complicated.
    \item At this point, if we draw many cells, as long as it is reasonably many, we will end up with a similar distribution for genes, since all the numbers so far are drawn from distributions.
    \item The reporter distribution essentially shifts the distribution to the "true distribution"  (this is a linear transformation on the mean of the distribution)
    \item This is repeated over $c$ cells to have a observed bin for a sample.  This is then repeated over all samples and all guides.
\end{itemize}
[FILL]

\subsection*{Multiplicity of Infection (MOI)}
\begin{itemize}
    \item Multiplicity of Infection (MOI) = the rate of virions to cells
    \item Effective coverage = the number of times we observe a guide
    \item The higher the MOI, the better.  A low MOI leads to low effective coverage - most cells are wasted because they were never genetically altered.
\end{itemize}
\begin{center}
    \includegraphics*[scale=0.8]{W9_7.png}
    \includegraphics*[scale=0.8]{W9_8.png}
\end{center}





\end{document}


